### PR TITLE
adds sloth holding

### PR DESCRIPTION
**Pull request by imsxz**
## About The Pull Request
Currently the ability to hold certain mobs is nonfunctional without admin intervention, along side a lot of bugs regarding picking up and holding said mobs, as well as a lot of hardcoding for which mobs can even be held and their in hand sprites. My aim is to fix the bugs that i'm aware of, as well as making it a lot easier for coders/admins to add and or modify what mobs can be held and what they look like. At the time of writing this, I've fixed the previous inability to scoop mobs up at all, putting this pull request up early to gather feedback and ideas regarding the direction I should be going in. Finally, I have at least 1 additional mob planned to allow players to hold, more if sprites are provided.

## Why It's Good For The Game
hold the pupper

## Changelog
:cl:imsxz and Twaticus
fix: you can now hold corgis and pAI's again. pull them, and then drag their sprite to yours.
add: you can now hold all of the stations pets, and more!
add: (b)admins can now make mobs holdable
add: you can now wash all of the stations pets in washing machines
add: pets can be put directly into washing machines from your hand
/:cl:
twaticus has helped a lot in identifying issues and testing this with me, as well as bringing this to my attention in the first place. 

todo:
[done] fix holding in general 
[done]add holding runtime the cat
[done]make mob holder sprites not hardcoded
[done]make creating new mob holders not snowflake trash
[done]^^double check to make sure these dont cause issues
harass kevinz on discord to relieve stress
[done] fix whatever this dumb runtime is Runtime in garbage.dm, line 322
[done]fix moving mob holders around your inventory
[done]fix deleting mob holders(via dropping and the likes)
[done]look into borgs wearing drones/pAI/etc
[not in the scope of this pr]make held mobs experience atmosphere effects
[fixed]pulling a mob out of a backpack that you're wearing releases them from their holder item, but if the backpack is on the ground it lets you equip it just fine.

future changes will probably be made by other people, im prone to forget stuff and get carried away with other ideas, but i had some that i wanted to add but would have been too much extra work for 1 PR.
planned:tm: future:tm: features and changes after the pr is merged include
more direct interactions with held mobs such as using items on them as if they were a normal mob
throwing held mobs properly
more interactions similar to the washing machine one, acting as an alternative to aggro grab moving mobs, such as depositing a held mob directly into the gibber
stuff suggested in the comments

### CODE REVIEW

**Comment by imsxz** [comment]
yes i know the current change is bad but it's a placeholder ok

File: code/modules/mob/living/living.dm [hunk 1]
```diff
@@ -1280,16 +1280,14 @@
 	..()
 	update_z(new_z)
 
-/mob/living/MouseDrop(mob/over)
+/mob/living/MouseDrop_T(atom/dropping, atom/user)
+	var/mob/living/U = user
```
Comment (bobbahbrown) [inline code/modules/mob/living/living.dm]: you may need to check if user is null when cast to /mob/living

Comment (bobbahbrown) [reply code/modules/mob/living/living.dm]: also may need to check for proximity

Comment (imsxz) [reply code/modules/mob/living/living.dm]: im not too familiar with mousedrop_t, how could user end up null there? i tried clickdragging mobs onto items/objects/etc and vise versa after reading this to be sure, no runtimes

Comment (bobbahbrown) [reply code/modules/mob/living/living.dm]: yes i pulled your branch and i agree it seems to be good, its not a problem from what i see its all good there

**Comment by bobbahbrown** [comment]
dropping ian / throwing ian causes runtime

image

**Comment by imsxz** [comment]
added to the todo list

**Comment by Spepbuthewranglesgithub** [comment]
this is a very important fix that is absolutely required for the server to properly continue


_oh god oh fuck the clown is carrying ian to an airlock somebody stop him_

File: code/modules/mob/living/living.dm (hunk 1 above)
Comment (Tlaltecuhtli) [reply code/modules/mob/living/living.dm]: then mob/living/MouseDrop_T(atom/dropping, mob/living//user) ?

**Comment by zxaber** [comment]
Does this fix putting drones/pAIs onto borg heads? Because if so, I'll remove them from the borg hat blacklist.

**Comment by Ziiro** [comment]
https://github.com/tgstation/tgstation/pull/39606

I think I had lizard/lizard pAI sprites added to my PR when I did it, if you want to raid it.

**Comment by imsxz** [comment]
> 
> 
> Does this fix putting drones/pAIs onto borg heads completely breaking? Because if so, I'll remove them from the borg hat blacklist.

if this issue is specific to borgs then i will put it on the todo list. currently trying to move them to different inventory slots makes the animal get released

**Comment by imsxz** [comment]
the runtime issues bobbah was referencing are now fixed, you can now put ian into your backpack. no other runtimes occurred when i abused my dog

**Comment by PKPenguin321** [comment]
Is that a child in your bag?

**Comment by Ziiro** [comment]
>now put ian into your backpack

Does being inside of a backpack call the proper turf atmos? i.e, can ian breathe if you can breathe? I recall this being a problem on my previous attempt.

**Comment by imsxz** [comment]
i ut ian into a bag of holding while i was wearing the debug outfit and went to space, he didnt die. another one for the list.

**Comment by imsxz** [comment]
(tested with just ian as of now, more tests to make sure it isnt fucked tomorrow)
newest commit makes any living mob holdable(though they cant be held without the variable for it set), as well as changing the sprite for held mobs much easier for admins and coders.

**Comment by Ziiro** [comment]
If you want a good testing implementation make it so getting dwarfed makes someone able to be picked up

Note this means you'll have to fun problems like:
* A mob that can be held suddenly becoming unhold-able
* A mob that can be held holding a mob that can be held holding a mob that can be held that...
* Deep frying mob holders
* Washing machine mob holders

**Comment by imsxz** [comment]
i wanted to look into adding holding dwarves after i checked out your PR(s), i think that would be best for a later PR, due to the issues that it opens up.

**Comment by Ziiro** [comment]
mobs holding mobs is indeed pandora's box

**Comment by zxaber** [comment]
I tested your branch in regards to drones on borg heads. Unfortunately, the drone mob holder object doesn't have `slot_flags`'s head slot enabled, so no one can wear it.

That aside, when I var edited it to allow for head placement, most of the issues I recall from before are fixed. The only remaining issue I noticed is that if the drone willingly leaves the borg, it doesn't update the borg's `hat` var or icon.

If you do end up re-adding the ability to place them in the head slot and wish to remove them from the borg blacklist, you can place this code into the robot.dm file to fix the borg issues;

```
/mob/living/silicon/robot/Exited(atom/A)
	if(hat && hat == A)
		hat = null
	update_icons()
	. = ..()
```

**Comment by imsxz** [comment]
planned on going over drones and pAI soon, finishing up the core of the changes today. also i took some sprites https://github.com/OracleStation/OracleStation/pull/907 . not sure who to credit for them, but i planned on having the animals which have sprites be holdable. @Ziiro

**Comment by Ziiro** [comment]
I actually think that it was Quantum-M who did those? It's been a while

**Comment by imsxz** [comment]
updated drone/pAI, havent tested it yet, they shouldnt need admin intervention to be put on borg heads anymore. last changes should be done within the next day or 2.

**Comment by trollbreeder** [comment]
Time to stuff 30 corgis into a BoH.

File: code/modules/mob/living/carbon/monkey/monkey.dm [hunk 2]
```diff
@@ -15,6 +15,10 @@
 	bodyparts = list(/obj/item/bodypart/chest/monkey, /obj/item/bodypart/head/monkey, /obj/item/bodypart/l_arm/monkey,
 					 /obj/item/bodypart/r_arm/monkey, /obj/item/bodypart/r_leg/monkey, /obj/item/bodypart/l_leg/monkey)
 	hud_type = /datum/hud/monkey
+	//can_be_held = TRUE //uncomment when someone wants to make sure monkeys cant steal shit while being held
+	//cam_head = TRUE
```
Comment (Ziiro) [inline code/modules/mob/living/carbon/monkey/monkey.dm]: `cam_head`

Comment (imsxz) [reply code/modules/mob/living/carbon/monkey/monkey.dm]: fuck i'll just remove that part

**Comment by imsxz** [comment]
ready to go, all of the main functionality works. remaining issues:
-if a container with a held mob is dumped onto the floor without going through a mobs hands, the mob isnt automatically released. not a problem for player controlled mobs, easy fix for NPC mobs(pick it up and drop it)
-the issue with borgs being able to wear drones that zxaber mentioned isnt resolved, but this isnt a new issue, and it will never come up because mob holders are blacklisted from borg heads.

**Comment by imsxz** [comment]
why the hell are people that can’t even watch the video allowed to like or dislike it that says not fucking FAIR FUCK YOUTUBE

**Comment by trollbreeder** [comment]
huh? why did you change the title?

**Comment by imsxz** [comment]
mom always said if you are trying to sell something you should appeal to the target audience

File: code/modules/mob/living/living.dm [hunk 3]
```diff
@@ -1280,18 +1280,19 @@
...
+	if(isliving(dropping))
+		var/mob/living/M = dropping
+		if(M.can_be_held && U.pulling == M)
+			M.mob_try_pickup(U)//blame kevinz
+			return//dont open the mobs inventory if you are picking them up
 	. = ..()
-	var/mob/living/user = usr
-	if(!istype(over) || !istype(user))
-		return
-	if(!over.Adjacent(src) || (user != src) || !canUseTopic(over))
```
Comment (AnturK) [inline code/modules/mob/living/living.dm]: Why did you remove these checks.

File: code/modules/mob/living/living_defines.dm [hunk 4]
```diff
@@ -110,6 +110,7 @@
 	var/list/obj/effect/proc_holder/abilities = list()
 
 	var/can_be_held = FALSE	//whether this can be picked up and held.
+	var/can_head = FALSE //if it can be held, can it fit on peoples heads?
```
Comment (AnturK) [inline code/modules/mob/living/living_defines.dm]: We're not short on letters, you could name this something specific.

Also probably better off to just make this generic designated slots instead. (carrying monkeys/sloths on your back or whatever)

File: code/game/machinery/washing_machine.dm [hunk 5]
```diff
@@ -239,6 +239,10 @@ GLOBAL_LIST_INIT(dye_registry, list(
 	addtimer(VARSET_CALLBACK(src, freshly_laundered, FALSE), 5 MINUTES, TIMER_UNIQUE | TIMER_OVERRIDE)
 	..()
 
+/obj/item/clothing/head/mob_holder/machine_wash(obj/machinery/washing_machine/WM)
```
Comment (AnturK) [inline code/game/machinery/washing_machine.dm]: This seems redundant if you automatically release it on putting it in?

**Review by AnturK — CHANGES_REQUESTED** [review]

**Comment by kriskog** [comment]
merge conflict

File: code/game/machinery/washing_machine.dm (hunk 5 above)
Comment (imsxz) [reply code/game/machinery/washing_machine.dm]: i decided to leave this in on the off chance that someone managed to stick a held ian into the washing machine without getting "attackby" called. i can remove it if you feel it's unnecessary.

File: code/modules/mob/living/living.dm (hunk 3 above)
Comment (imsxz) [reply code/modules/mob/living/living.dm]: they are handled in mousedrop_t and the pulling check

**Comment by imsxz** [comment]
why would travis do this to me

**Comment by kriskog** [comment]
Because
```
file ./code/modules/mob/living/living_defines.dm is missing a trailing newline
```

File: code/modules/mob/living/living.dm (hunk 3 above)
Comment (AnturK) [reply code/modules/mob/living/living.dm]: I don't follow.

File: code/game/machinery/washing_machine.dm (hunk 5 above)
Comment (AnturK) [reply code/game/machinery/washing_machine.dm]: Please do, and if you're worried about weird stuff, you can unload these in /Entered instead.

File: code/modules/mob/living/living.dm (hunk 3 above)
Comment (imsxz) [reply code/modules/mob/living/living.dm]: all of the checks that i removed there are already handled by the checks in mousedrop and the ability to drag someone.

**Comment by imsxz** [comment]
i did some testing and it was the attackby that was redundant from past testing, ive opted to keep the machine_wash instead, works fine. this is otherwise ready to go.

**Comment by imsxz** [comment]
im ognna scream

**Comment by imsxz** [comment]
ok i apologize for the trouble i kept thinking i had finished each commit ive done today
